cmd/thanos/receive.go

```diff
@@ -286,7 +286,7 @@ func runReceive(
 					}
 					level.Info(logger).Log("msg", "tsdb started")
 					localStorage.Set(db.Get(), startTimeMargin)
-					webHandler.SetWriter(receive.NewWriter(log.With(logger, "component", "receive-writer"), localStorage))
+					webHandler.SetWriter(receive.NewWriter(log.With(logger, "component", "receive-writer"), reg, localStorage))
```

Comment:
Have you tested this on a receive that flushes it’s database? I don’t think this should work. Receive creates a new writer, and thus would register metrics, every time every time it flushes it’s DB. But if you try to register the same metrics multiple times, the registerer will panic. I think you will need to use the receive.UnRegisterer here.